src/lib.rs

```diff
@@ -35,39 +34,210 @@ thread_local! {
 
 /// UserdataGuard pops an entry off the USERDATA stack, restoring the
 /// thread-local state to its value previous to the creation of the UserdataGuard.
-pub struct UserdataGuard();
-pub struct UserdataError();
+/// Invariants: As long as a UserdataGuard is live:
+//// - The stack of userdata items for this thread must have at least one item.
+///  - The top item on that stack must be the one this guard was built with.
+///  - The `data` field must not be None.
+/// If any of these invariants fails, try_drop will return an error.
+pub struct UserdataGuard {
+    // Keep a copy of the data we expect to be popping off the stack. This allows
+    // us to check for consistency, and also serves to make this type !Send:
+    // https://doc.rust-lang.org/nightly/std/primitive.pointer.html#impl-Send-1
+    data: Option<*mut c_void>,
+}
 
-impl Drop for UserdataGuard {
-    fn drop(&mut self) {
+impl UserdataGuard {
+    fn new(u: *mut c_void) -> Self {
+        UserdataGuard { data: Some(u) }
+    }
+
+    /// Even though we have a Drop impl on this guard, when possible it's
+    /// best to call try_drop explicitly. That way any failures of internal
+    /// variants can be signaled to the user immediately by returning
+    /// rustls_result::Panic.
+    fn try_drop(mut self) -> Result<(), UserdataError> {
+        self.try_pop()
+    }
+
+    fn try_pop(&mut self) -> Result<(), UserdataError> {
+        let expected_data = match self.data {
+            Some(d) => d,
+            None => return Err(UserdataError::AlreadyPopped),
+        };
+        eprintln!("popping");
         USERDATA
             .try_with(|userdata| {
-                if let Ok(mut v) = userdata.try_borrow_mut() {
-                    v.pop();
-                }
+                userdata.try_borrow_mut().map_or_else(
+                    |_| Err(UserdataError::AlreadyBorrowed),
+                    |mut v| match v.pop() {
```

Comment:
And here... If you combine it with `?`, you can reduce the rightward drift a fair bit.